### PR TITLE
Added Knuth–Morris–Pratt in Python3 

### LINKED ISSUE #691: Knuth–Morris–Pratt in Python3
This is an open issue for the Knuth–Morris–Pratt algorithm implementation in Python3. If you wish to contribute, just make some noise and we will assign you.

### PR DESCRIPTION
**Pull request by riturajjain2000**
**Fixes #691**

By submitting this pull request I confirm I've read and complied with the below declarations.

-   [x] I have read the [Contribution guidelines](https://github.com/FOSS-UCSC/FOSSALGO/blob/master/CONTRIBUTING.md) 
and I am confident that my PR reflects them.

-   [x] I have followed the [coding guidelines](https://github.com/FOSS-UCSC/FOSSALGO/blob/master/CONTRIBUTING.md#cs) 
for this project.

-   [x] My code follows the [skeleton code structure](https://github.com/FOSS-UCSC/FOSSALGO/blob/master/CONTRIBUTING.md#sample).

-   [x] This pull request has a descriptive title. 
For example, `Added {Algorithm/DS name} [{Language}]`, not `Update README.md` or `Added new code`.

-   [x] This pull request shall only be merged once the build passes.

-   [x] I have mentioned the issue number correctly (with hyper-link) in this pull request description.

### CODE REVIEW

File: algorithms/ar-kmp/python3/knuth_morris_pratt.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# Python program for KMP Algorithm 
+
+def LPSArray(pat, M, lps): 
+	len = 0 
+
+	lps[0]
```
Comment (RavinduSachintha) [inline algorithms/ar-kmp/python3/knuth_morris_pratt.py:6]: Codacy found an issue: [Statement seems to have no effect](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6363416)

File: algorithms/ar-kmp/python3/knuth_morris_pratt.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+	# create lps[] that will hold the longest prefix suffix values for pattern 
+	lps = [0]*M 
+	j = 0 
+
+	# Preprocess the pattern (calculate lps[] array) 
+	LPSArray(pat, M, lps) 
+
+	i = 0 # index for txt[] 
+	while i < N: 
+		if pat[j] == txt[i]:
```
Comment (RavinduSachintha) [inline algorithms/ar-kmp/python3/knuth_morris_pratt.py:37]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6363416)

File: algorithms/ar-kmp/python3/knuth_morris_pratt.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+				lps[i] = 0
+				i += 1
+
+def KMP(pat, txt): 
+	M = len(pat) 
+	N = len(txt) 
+
+	# create lps[] that will hold the longest prefix suffix values for pattern 
+	lps = [0]*M 
+	j = 0
```
Comment (RavinduSachintha) [inline algorithms/ar-kmp/python3/knuth_morris_pratt.py:30]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6363416)

File: algorithms/ar-kmp/python3/knuth_morris_pratt.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+				
+			else: 
+				lps[i] = 0
+				i += 1
+
+def KMP(pat, txt): 
+	M = len(pat) 
+	N = len(txt) 
+
+	# create lps[] that will hold the longest prefix suffix values for pattern
```
Comment (RavinduSachintha) [inline algorithms/ar-kmp/python3/knuth_morris_pratt.py:28]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6363416)

File: algorithms/ar-kmp/python3/knuth_morris_pratt.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
+# Python program for KMP Algorithm 
+
+def LPSArray(pat, M, lps): 
+	len = 0
```
Comment (RavinduSachintha) [inline algorithms/ar-kmp/python3/knuth_morris_pratt.py:4]: Codacy found an issue: [Redefining built-in 'len'](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6363416)

File: algorithms/ar-kmp/python3/knuth_morris_pratt.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+			if j != 0: 
+				j = lps[j-1] 
+			else: 
+				i += 1
+
+
+
+txt = "ABABDABACDABABCABAB"
+pat = "ABABCABAB"
+KMP(pat, txt)
```
Comment (RavinduSachintha) [inline algorithms/ar-kmp/python3/knuth_morris_pratt.py:56]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6363416)

File: algorithms/ar-kmp/python3/knuth_morris_pratt.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+	N = len(txt) 
+
+	# create lps[] that will hold the longest prefix suffix values for pattern 
+	lps = [0]*M 
+	j = 0 
+
+	# Preprocess the pattern (calculate lps[] array) 
+	LPSArray(pat, M, lps) 
+
+	i = 0 # index for txt[]
```
Comment (RavinduSachintha) [inline algorithms/ar-kmp/python3/knuth_morris_pratt.py:35]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6363416)

File: algorithms/ar-kmp/python3/knuth_morris_pratt.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+			if len != 0: 
+				len = lps[len-1] 
+
+				
+			else: 
+				lps[i] = 0
+				i += 1
+
+def KMP(pat, txt): 
+	M = len(pat)
```
Comment (RavinduSachintha) [inline algorithms/ar-kmp/python3/knuth_morris_pratt.py:25]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6363416)

File: algorithms/ar-kmp/python3/knuth_morris_pratt.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
+# Python program for KMP Algorithm 
+
+def LPSArray(pat, M, lps):
```
Comment (RavinduSachintha) [inline algorithms/ar-kmp/python3/knuth_morris_pratt.py:3]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6363416)

File: algorithms/ar-kmp/python3/knuth_morris_pratt.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+		else: 
+			if len != 0: 
+				len = lps[len-1] 
+
+				
+			else: 
+				lps[i] = 0
+				i += 1
+
+def KMP(pat, txt):
```
Comment (RavinduSachintha) [inline algorithms/ar-kmp/python3/knuth_morris_pratt.py:24]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6363416)

File: algorithms/ar-kmp/python3/knuth_morris_pratt.py [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+	i = 1
+
+	
+	while i < M: 
+		if pat[i]== pat[len]: 
+			len += 1
+			lps[i] = len
+			i += 1
+		else: 
+			if len != 0:
```
Comment (RavinduSachintha) [inline algorithms/ar-kmp/python3/knuth_morris_pratt.py:16]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6363416)

File: algorithms/ar-kmp/python3/knuth_morris_pratt.py [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+			else: 
+				lps[i] = 0
+				i += 1
+
+def KMP(pat, txt): 
+	M = len(pat) 
+	N = len(txt) 
+
+	# create lps[] that will hold the longest prefix suffix values for pattern 
+	lps = [0]*M
```
Comment (RavinduSachintha) [inline algorithms/ar-kmp/python3/knuth_morris_pratt.py:29]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6363416)

File: algorithms/ar-kmp/python3/knuth_morris_pratt.py [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+	lps[0]
+	i = 1
+
+	
+	while i < M: 
+		if pat[i]== pat[len]: 
+			len += 1
+			lps[i] = len
+			i += 1
+		else:
```
Comment (RavinduSachintha) [inline algorithms/ar-kmp/python3/knuth_morris_pratt.py:15]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6363416)

File: algorithms/ar-kmp/python3/knuth_morris_pratt.py [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+	LPSArray(pat, M, lps) 
+
+	i = 0 # index for txt[] 
+	while i < N: 
+		if pat[j] == txt[i]: 
+			i += 1
+			j += 1
+
+		if j == M: 
+			print ("Found pattern at index " + str(i-j))
```
Comment (RavinduSachintha) [inline algorithms/ar-kmp/python3/knuth_morris_pratt.py:42]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6363416)

File: algorithms/ar-kmp/python3/knuth_morris_pratt.py [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+
+	i = 0 # index for txt[] 
+	while i < N: 
+		if pat[j] == txt[i]: 
+			i += 1
+			j += 1
+
+		if j == M: 
+			print ("Found pattern at index " + str(i-j)) 
+			j = lps[j-1]
```
Comment (RavinduSachintha) [inline algorithms/ar-kmp/python3/knuth_morris_pratt.py:43]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6363416)

File: algorithms/ar-kmp/python3/knuth_morris_pratt.py [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+def KMP(pat, txt): 
+	M = len(pat) 
+	N = len(txt) 
+
+	# create lps[] that will hold the longest prefix suffix values for pattern 
+	lps = [0]*M 
+	j = 0 
+
+	# Preprocess the pattern (calculate lps[] array) 
+	LPSArray(pat, M, lps)
```
Comment (RavinduSachintha) [inline algorithms/ar-kmp/python3/knuth_morris_pratt.py:33]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6363416)

File: algorithms/ar-kmp/python3/knuth_morris_pratt.py [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+			i += 1
+			j += 1
+
+		if j == M: 
+			print ("Found pattern at index " + str(i-j)) 
+			j = lps[j-1] 
+
+		# mismatch after j matches 
+		elif i < N and pat[j] != txt[i]: 
+			if j != 0:
```
Comment (RavinduSachintha) [inline algorithms/ar-kmp/python3/knuth_morris_pratt.py:47]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6363416)

File: algorithms/ar-kmp/python3/knuth_morris_pratt.py [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+		if pat[j] == txt[i]: 
+			i += 1
+			j += 1
+
+		if j == M: 
+			print ("Found pattern at index " + str(i-j)) 
+			j = lps[j-1] 
+
+		# mismatch after j matches 
+		elif i < N and pat[j] != txt[i]:
```
Comment (RavinduSachintha) [inline algorithms/ar-kmp/python3/knuth_morris_pratt.py:46]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6363416)

File: algorithms/ar-kmp/python3/knuth_morris_pratt.py (hunk 5 above)
Comment (RavinduSachintha) [inline algorithms/ar-kmp/python3/knuth_morris_pratt.py:4]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6363416)

File: algorithms/ar-kmp/python3/knuth_morris_pratt.py [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+
+	# create lps[] that will hold the longest prefix suffix values for pattern 
+	lps = [0]*M 
+	j = 0 
+
+	# Preprocess the pattern (calculate lps[] array) 
+	LPSArray(pat, M, lps) 
+
+	i = 0 # index for txt[] 
+	while i < N:
```
Comment (RavinduSachintha) [inline algorithms/ar-kmp/python3/knuth_morris_pratt.py:36]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6363416)

File: algorithms/ar-kmp/python3/knuth_morris_pratt.py [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+
+	
+	while i < M: 
+		if pat[i]== pat[len]: 
+			len += 1
+			lps[i] = len
+			i += 1
+		else: 
+			if len != 0: 
+				len = lps[len-1]
```
Comment (RavinduSachintha) [inline algorithms/ar-kmp/python3/knuth_morris_pratt.py:17]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6363416)

File: algorithms/ar-kmp/python3/knuth_morris_pratt.py [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+	# Preprocess the pattern (calculate lps[] array) 
+	LPSArray(pat, M, lps) 
+
+	i = 0 # index for txt[] 
+	while i < N: 
+		if pat[j] == txt[i]: 
+			i += 1
+			j += 1
+
+		if j == M:
```
Comment (RavinduSachintha) [inline algorithms/ar-kmp/python3/knuth_morris_pratt.py:41]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6363416)

File: algorithms/ar-kmp/python3/knuth_morris_pratt.py [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+	while i < N: 
+		if pat[j] == txt[i]: 
+			i += 1
+			j += 1
+
+		if j == M: 
+			print ("Found pattern at index " + str(i-j)) 
+			j = lps[j-1] 
+
+		# mismatch after j matches
```
Comment (RavinduSachintha) [inline algorithms/ar-kmp/python3/knuth_morris_pratt.py:45]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6363416)

File: algorithms/ar-kmp/python3/knuth_morris_pratt.py [hunk 23]
```diff
@@ -0,0 +1,58 @@
+# Python program for KMP Algorithm
```
Comment (RavinduSachintha) [inline algorithms/ar-kmp/python3/knuth_morris_pratt.py:1]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6363416)

File: algorithms/ar-kmp/python3/knuth_morris_pratt.py [hunk 24]
```diff
@@ -0,0 +1,58 @@
+# Python program for KMP Algorithm 
+
+def LPSArray(pat, M, lps): 
+	len = 0 
+
+	lps[0]
+	i = 1
+
+
```
Comment (RavinduSachintha) [inline algorithms/ar-kmp/python3/knuth_morris_pratt.py:9]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6363416)

File: algorithms/ar-kmp/python3/knuth_morris_pratt.py [hunk 25]
```diff
@@ -0,0 +1,49 @@
...
+	lenn = 0
+	i = 1
+	while i < M:
+		if pat[i]== pat[lenn]:
+			lenn += 1
+			lps[i] = lenn
+			i += 1
+		else:
+			if lenn != 0:
+				lenn = lps[lenn-1]
```
Comment (RavinduSachintha) [inline algorithms/ar-kmp/python3/knuth_morris_pratt.py:12]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6363416)

**Review by Danushka96 — APPROVED** [review]

**Pull request merged**